modules/frontend/tag_handlers.go

```diff
@@ -121,27 +122,149 @@ func streamingTags[TReq proto.Message, TResp proto.Message](ctx context.Context,
+	})
+}
+
+func newTagValuesHTTPHandler(cfg Config, next pipeline.AsyncRoundTripper[combiner.PipelineResponse], o overrides.Interface, logger log.Logger) http.RoundTripper {
+	postSLOHook := metadataSLOPostHook(cfg.Search.MetadataSLO)
+	return RoundTripperFunc(func(req *http.Request) (*http.Response, error) {
+		// if error is not nil, return error Response but suppress the error
+		tenant, errResp, err := extractTenantWithError(req, logger)
+		if err != nil {
+			return errResp, nil
+		}
+
+		// build and use round tripper
+		comb := combiner.NewTypedSearchTagValues(o.MaxBytesPerTagValuesQuery(tenant))
+		rt := pipeline.NewHTTPCollector(next, cfg.ResponseConsumers, comb)
+		start := time.Now()
+
+		resp, err := rt.RoundTrip(req)
+		// ask for the typed diff and use that for the SLO hook. it will have up to date metrics
+		var bytesProcessed uint64
+		searchResp, _ := comb.GRPCDiff()
+		if searchResp != nil && searchResp.Metrics != nil {
+			bytesProcessed = searchResp.Metrics.InspectedBytes
+		}
+
+		duration := time.Since(start)
+		postSLOHook(resp, tenant, bytesProcessed, duration, err)
+		logHTTPResult(logger, tenant, req.URL.Path, duration.Seconds(), err)
+
+		return resp, err
 	})
 }
 
+func newTagValuesV2HTTPHandler(cfg Config, next pipeline.AsyncRoundTripper[combiner.PipelineResponse], o overrides.Interface, logger log.Logger) http.RoundTripper {
+	postSLOHook := metadataSLOPostHook(cfg.Search.MetadataSLO)
+	return RoundTripperFunc(func(req *http.Request) (*http.Response, error) {
+		// if error is not nil, return error Response but suppress the error
+		tenant, errResp, err := extractTenantWithError(req, logger)
+		if err != nil {
+			return errResp, nil
+		}
+
+		// build and use round tripper
+		comb := combiner.NewTypedSearchTagValuesV2(o.MaxBytesPerTagValuesQuery(tenant))
+		rt := pipeline.NewHTTPCollector(next, cfg.ResponseConsumers, comb)
+		start := time.Now()
+
+		resp, err := rt.RoundTrip(req)
+
+		// ask for the typed diff and use that for the SLO hook. it will have up to date metrics
+		var bytesProcessed uint64
+		searchResp, _ := comb.GRPCDiff()
+		if searchResp != nil && searchResp.Metrics != nil {
+			bytesProcessed = searchResp.Metrics.InspectedBytes
+		}
+
+		duration := time.Since(start)
+		postSLOHook(resp, tenant, bytesProcessed, duration, err)
+
+		logHTTPResult(logger, tenant, req.URL.Path, duration.Seconds(), err)
+		return resp, err
+	})
+}
+
+func extractTenantWithError(req *http.Request, logger log.Logger) (string, *http.Response, error) {
+	tenant, err := user.ExtractOrgID(req.Context())
+	if err != nil {
+		level.Error(logger).Log("msg", "tags failed to extract orgid", "err", err)
+		return "", &http.Response{
+			StatusCode: http.StatusBadRequest,
+			Status:     http.StatusText(http.StatusBadRequest),
+			Body:       io.NopCloser(strings.NewReader(err.Error())),
+		}, err
+	}
+	return tenant, nil, err
+}
+
+func logHTTPResult(logger log.Logger, tenantID, path string, durationSeconds float64, err error) {
```

Comment:
Since this method is reused for several things, it might make sense to change the message `search tags response` to be more generic like `metadata response`, or parameterize it.